File: Data Structures & Algorithms/majority-element/submission-1.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        mp : dict[int, int] = {}
        size = len(nums)

        for num in nums : 
            mp[num] = mp.get(num,0) + 1
            if mp.get(num,0) > size/2 :
                return num

        return -1 

File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Count frequency of each number
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }

        // Use a min-heap to keep top k frequent elements
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;

        for (auto& entry : freq) {
            minHeap.push({entry.second, entry.first});  // {frequency, number}
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }

        // Extract the top k elements from the heap
        vector<int> result;
        while (!minHeap.empty()) {
            result.push_back(minHeap.top().second);
            minHeap.pop();
        }

        return result;
    }
};
